### PR TITLE
Agent Retriever Updates

### LINKED ISSUE #40: Add refactor for scraping manpages
See vitor comments on : https://github.com/The-OpenROAD-Project/ORAssistant/pull/31

### PR DESCRIPTION
**Pull request by palaniappan-r**
- Add support for adding PDF files to the vector database
- Add an OpenSTA retriever node in the graph workflow
- Rewrite scraper script to build docs from OR/ORFS github repos
- Rewrite tool descriptions and llm prompts

Resolves #40 and #37 and #36 and #32

### CODE REVIEW

File: backend/Dockerfile [hunk 1]
```diff
@@ -9,9 +9,9 @@ RUN pip install --upgrade -r /ORAssistant-backend/requirements.txt
 COPY . .
 
 # Prereqs
-RUN apt-get update && apt-get install -y pandoc git
+RUN apt-get update && apt-get install -y pandoc git make
```
Comment (luarss) [inline backend/Dockerfile:12]: Double check if the base image has `make`

File: backend/build_docs.py [hunk 2]
```diff
@@ -0,0 +1,210 @@
...
+def update_source_dict(dir_path: str) -> None:
+    print(f'Updating source_dict for {dir_path}...')
+    for root, _, files in os.walk(dir_path):
+        for file in files:
+            file_path = os.path.join(root, file)
+            file_name = os.path.basename(file_path)
+            source_dict[file_path] = file_name
+
+
+def clone_repo(url: str, folder_name: str) -> None:
```
Comment (luarss) [inline backend/build_docs.py:30]: Can you try to integrate the OR_docs and manpage together? At least tie them to the same commit hash number

File: backend/build_docs.py [hunk 3]
```diff
@@ -0,0 +1,210 @@
...
+    else:
+        print('Failed to download file. Status code:', response.status_code)
+
+
+if __name__ == '__main__':
+    docs_paths = ['data']
+    purge_folders(folder_paths=docs_paths)
+
+    clone_repo(
+        url='https://github.com/The-OpenROAD-Project/OpenROAD.git',
```
Comment (luarss) [inline backend/build_docs.py:175]: Lock a hash
This should work: https://stackoverflow.com/questions/75454944/how-to-clone-a-repository-from-a-specific-commit-id

File: backend/requirements.txt [hunk 4]
```diff
@@ -4,8 +4,15 @@ langchain-huggingface==0.0.3
...
 markdown==3.0.0
 rank-bm25==0.2.2
+sphinx==7.4.4
+sphinx-autobuild==2024.4.16
+myst-parser==3.0.1
+sphinxcontrib-mermaid==0.9.2
+sphinx-book-theme==1.1.3
+sphinx-copybutton==0.5.2
+sphinx-external-toc==1.0.1
+pypdf==4.2.0
```
Comment (luarss) [inline backend/requirements.txt:18]: is Pypdf accurate? How does it compare to this text-based file

https://github.com/The-OpenROAD-Project/OpenSTA/blob/master/doc/OpenSTA.odt

File: backend/src/api/routers/graphs.py [hunk 5]
```diff
@@ -81,10 +81,40 @@ async def get_agent_response(user_input: UserInput) -> dict[str, Union[str, list
...
+        and len(output) > 0
+        and 'agent' in output[0]
+        and 'tools' in output[0]['agent']
+        and len(output[0]['agent']['tools']) > 0
+    ):
+        tool = output[0]['agent']['tools'][0].get('name', '')
+    else:
+        print('Tool name extraction failed')
+
+    if tool and isinstance(output, list) and len(output) > 1 and tool in output[1]:
```
Comment (luarss) [inline backend/src/api/routers/graphs.py:95]: Is it possible for `tool` to be a empty string? Or in other words, is it possible for `output[0]['agent']['tools'][0]` dict to have no `tool` key?

File: backend/src/api/routers/graphs.py [hunk 6]
```diff
@@ -81,10 +81,40 @@ async def get_agent_response(user_input: UserInput) -> dict[str, Union[str, list
...
+        if 'context' in output[1][tool]:
+            context = output[1][tool].get('context', '')
+        else:
+            print('Context extraction failed')
+
+        if 'sources' in output[1][tool]:
+            sources = output[1][tool].get('sources', '')
+        else:
+            print('Sources extraction failed')
```
Comment (luarss) [inline backend/src/api/routers/graphs.py:104]: Is it fatal for the code if any source / context extraction failed?

File: backend/src/chains/multi_retriever_chain.py [hunk 7]
```diff
@@ -20,17 +20,18 @@ def __init__(
...
         docs_path: Optional[list[str]] = None,
         manpages_path: Optional[list[str]] = None,
-        embeddings_model_name: str = 'BAAI/bge-large-en-v1.5',
+        pdfs_path: Optional[list[str]] = None,
+        embeddings_model_name: Optional[str] = None,
         use_cuda: bool = False,
-        search_k: list[int] = [5, 5],
-        weights: list[float] = [0.5, 0.5],
+        search_k: list[int] = [5, 5, 5],
+        weights: list[float] = [0.4, 0.4, 0.1],
```
Comment (luarss) [inline backend/src/chains/multi_retriever_chain.py:27]: Why do they not sum up to 1.0?

**Comment by luarss** [comment]
@palaniappan-r Tip: Do use github magic keywords if the PR solves an existing issue 
https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue

File: backend/src/api/routers/graphs.py (hunk 6 above)
Comment (palaniappan-r) [reply backend/src/api/routers/graphs.py:104]: No, the sources/context sent back in the response body will be empty in that case.

File: backend/src/api/routers/graphs.py (hunk 5 above)
Comment (palaniappan-r) [reply backend/src/api/routers/graphs.py:95]: Not anymore, this snippet was written to help with debugging.

**Comment by palaniappan-r** [comment]
> @palaniappan-r Tip: Do use github magic keywords if the PR solves an existing issue https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue

Done, please check now.

**Review by luarss — APPROVED** [review]

**Pull request merged**